enhancements/api-review/azure_user_defined_tags.md

```diff
@@ -0,0 +1,322 @@
+                      type: string
+                    description: UserTags additional keys and values that the installer
+                      will add as tags to all resources that it creates. Resources
+                      created by the cluster itself may not include these tags.
+                  type: object
+```
+
+- Add `resourceTags` field to `status.platformStatus.azure` 
+  of the `infrastructure.config.openshift.io`
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: infrastructures.config.openshift.io
+spec:
+  versions:
+  - name: v1
+    schema:
+      openAPIV3Schema:
+        properties:
+          status:
+            properties:
+              platformStatus:
+                properties:
+                  azure:
+                    properties:
+                      resourceTags:
+                        description: resourceTags is a list of additional tags to apply to Azure
+                        resources created for the cluster. See 
+                        https://docs.microsoft.com/en-us/rest/api/resources/tags for information 
+                        on tagging Azure resources. Azure supports a maximum of 50 tags per 
+                        resource except for few, which have limitation of 15 tags. OpenShift 
+                        reserves 10 tags for its internal use, and allows 5 tags 
+                        for user configuration.
+                        type: array
+                        maxItems: 5
+                        items:
+                          description: AzureResourceTag is a tag to apply to Azure resources created for the cluster.
+                          type: object
+                          required:
+                            - key
+                            - value
+                          properties:
+                            key:
+                              description: key is the key of the tag
+                              type: string
+                              maxLength: 128
+                              minLength: 1
+                              pattern: ^[a-zA-Z][0-9A-Za-z_.=+-@]+$
+                            value:
+                              description: value is the value of the tag.
+                              type: string
+                              maxLength: 256
+                              minLength: 1
+                              pattern: ^[0-9A-Za-z_.=+-@]+$
+```
+
+### Implementation Details/Notes/Constraints [optional]
+Add a new field `resourceTags` to `.status.platformStatus.azure` of the 
+`infrastructure.config.openshift.io` type. Tags included in the `resourceTags` field 
+will be applied to new resources created for the cluster by the in-cluster operators.
+
+The `resourceTags` field in `status.platformStatus.azure` of `infrastructure.config.openshift.io`
+will be populated by the installer using the entries from `platform.azure.userTags` field of `install-config`.
```

Comment:
does anything else go into resourceTags other than the user tags?  (I guess the `kubernetes.io.cluster:owned` tag would be one)

Do other tags(not from spec.userTags) go into resourceTags in the AWS equivalent?

Reply:
No, currently just the tags in userTags will be added to resourceTags. Behavior is same for AWS too.

Reply:
so we don't even put `kubernetes.io.cluster:owned` in there?  Is that just hardcoded into the resource controllers to apply?

Reply:
Yes, resourceTags will have only the user defined tags and the operators will add the default tag.

Reply:
> Yes, resourceTags will have only the user defined tags and the operators will add the default tag.

that seems a little strange to be honest.  I would have expected the user to supply their desired tags in spec, and then status to reflect all the user tags plus all the "default/required by ocp" tags.  And then the operators would look at status and apply that set of tags to the resources they manage.

Having each operator hardcode the default tag seems undesirably distributed(and has potential to be out of sync if one operator doesn't hardcode the right thing, especially if we change/add a new one) instead of centralized.

That said, i realize you're only building on how it works on AWS today, so unless we're going to revisit the AWS version of this implementation, it is probably moot.  I wouldn't want to do something different for Azure than we do for AWS.

Reply:
Yeah I proposed the implementation the way it was done for AWS. And also I think since userTags cannot be updated it was added in status, as keeping it in spec would give out the idea it can be modified.

Reply:
keeping it in status is fine, but you could(well, could have) still add the additional system-generated tag in status.

Reply:
Yeah, I agree. Will update the description of resourceTags to state it contains both OCP and user defined tags applied on the resources. 
Also, in a case where operator adds specific tag to a resource it manages, should it also update the resourceTags(it could be added to just one type of resource)?

Reply:
> Will update the description of resourceTags to state it contains both OCP and user defined tags applied on the resources.

so on AWS status won't include it, but on Azure it will?  And then operator/controller implementations will need to know whether or not to add the OCP tags to the list?

I started out asking the question because i wasn't sure what we did in AWS.  Much as i think it would be better to include it all in status, i also think it's probably better to be consistent in how we use status for this across our clouds.

So if we're not going to revisit the AWS status field, we should probably make the Azure one work the same way.

Reply:
Yeah, but since we are adding support now for azure and gcp, should we make this change in these both and later pick up aws? Otherwise, we will have to revisit all 3 in the future.